### PR TITLE
Refactor attention.py part 2

### PR DESCRIPTION
**Pull request by KshitijLakhani**
# Description

Part 2 of the attention refactoring in transformer_engine/pytorch directory. This PR addresses most of the TODOs in Part 1 and also the directory restructuring. It also removes the redundant `check_set_window_size` in `MultiHeadAttention` and `TransformerLayer`.
Refer to [Part 1](https://github.com/NVIDIA/TransformerEngine/pull/1542) for more info.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

## Changes

1. Finalized directory structure after discussions with @cyanguwa and @ptrendx :
```
te/pytorch:
  - transformer.py: TELayer
  - attention/
     - multihead_attention.py: MHA/GQA/MLA
     - dot_product_attention/
           - dot_product_attention.py: DPA 
           - backends.py: FusedAttention, FlashAttention, UnfusedDPA 
           - context_parallel.py: P2P, A2A, AllGather, attn_with_cp 
           - softmax.py: 
           - utils.py: 
     - rope.py
     - inference.py
```

2. `multi_head_attention.py`, `context_parallel.py` and `backends.py` were created while moving some more generic pytorch utils functions to `pytorch/utils.py` and also moving whatever is leftover of `attention.py` into `dot_product_attention.py`
3. Ensure that user code importing attention, for e.g., `import transformer_engine.pytorch.attention` does not break and this can still be used to import `DotProductAttention` and `MultiheadAttention` 

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/distributed.py [hunk 1]
```diff
@@ -1469,12 +1471,14 @@ def _fsdp_gather_tensors(
 
 def _is_te_module(module):
     """
-    Check if given module is a Transformer Engine module that requires the TE checkpoint
+    Check if given module is a Transsformer Engine module that requires the TE checkpoint
```
Comment (ksivaman) [inline transformer_engine/pytorch/distributed.py:1474]:
```suggestion
    Check if given module is a Transformer Engine module that requires the TE checkpoint
```

File: transformer_engine/pytorch/attention/__init__.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2022-2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+"""Python interface for attention"""
+
+from .dot_product_attention import DotProductAttention
+from .multi_head_attention import MultiheadAttention
+
+__all__ = ["DotProductAttention", "MultiheadAttention"]
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/__init__.py:10]: What about `InferenceParams`?

Comment (KshitijLakhani) [reply transformer_engine/pytorch/attention/__init__.py:10]: @cyanguwa I believe before the refactoring as well, `InferenceParams` was importable only when `import transformer_engine.pytorch` was performed and not when `import transformer_engine.pytorch.attention` was performed, which is still true in this PR as `transformer_engine/pytorch/__init__.py` does still allow the import of `InferenceParams`.
Within TE, the `InferenceParams` were accessed using the module that contained them, which is still the case with this PR. For e.g. `from transformer_engine.pytorch.attention.inference import InferenceParams`

The purpose of this file, `transformer_engine/pytorch/attention/__init__.py` was to not break the code for users who were doing a `import transformer_engine.pytorch.attention` /, `from transformer_engine.pytorch.attention import DotProductAttention`, etc. (which exposed `DotProductAttention` and `MultiheadAttention` only, before this PR and, that behavior continues)

However, if we wish to expose `InferenceParams` or if I may have misunderstood it's usage in the past, I can always add it to `InferenceParams`

File: transformer_engine/pytorch/transformer.py [hunk 3]
```diff
@@ -12,11 +12,9 @@
...
 from transformer_engine.pytorch.module import LayerNormMLP, LayerNorm, RMSNorm
 from transformer_engine.debug.pytorch.debug_state import TEDebugState
-from transformer_engine.pytorch.attention import (
-    MultiheadAttention,
-)
-from transformer_engine.pytorch.dot_product_attention.inference import InferenceParams
-from transformer_engine.pytorch.dot_product_attention.utils import check_set_window_size
+from transformer_engine.pytorch.attention.multi_head_attention import MultiheadAttention
+from transformer_engine.pytorch.attention.inference import InferenceParams
+from transformer_engine.pytorch.attention.dot_product_attention.utils import check_set_window_size
```
Comment (cyanguwa) [inline transformer_engine/pytorch/transformer.py:17]: Did you check if we could remove the potentially duplicate `check_set_window_size` in either `TransformerLayer` or `MultiHeadAttention`?

File: transformer_engine/pytorch/attention/dot_product_attention/backends.py [hunk 4]
```diff
@@ -0,0 +1,1619 @@
...
+                total_tokens,
+            )
+
+            # [tq, np, hn] --> [tq, hp]
+            context_layer = context_layer.view(total_tokens, -1)
+
+        return context_layer
+
+
+class FusedAttention(torch.nn.Module):
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/dot_product_attention/backends.py:376]: Would it be better if we keep the original order, `UnfusedDotProductAttention`, `_PrepareQKVForFA`, `FlashAttention`, `FusedAttnFunc`, `FusedAttention`? Long funcs/classes at the end.

File: transformer_engine/pytorch/attention/__init__.py (hunk 2 above)
Comment (cyanguwa) [reply transformer_engine/pytorch/attention/__init__.py:10]: I see - you are strictly following our old structure - thanks for doing that! Given that `inference.py` is in `attention/`, though, I wonder if it'd be better to really just add `from transformer_engine.pytorch.attention.inference import InferenceParams` to `attention/__init__.py`? Any side effects can you think of for this? Thanks!

Comment (cyanguwa) [reply transformer_engine/pytorch/attention/__init__.py:10]: I think we should add `RotaryPositionEmbedding` to both places as well, because users might need to set up RoPE outside their model like [here](https://github.com/NVIDIA/TransformerEngine/blob/e5a673f654d060b79decd6cef4fbfdddb0293c9b/docs/examples/te_llama/te_llama.py#L14).

**Comment by cyanguwa** [comment]
Please run L1 and L3 CI as well when you do the CI, because we have made a lot of changes. Thanks! :)

File: transformer_engine/pytorch/attention/__init__.py (hunk 2 above)
Comment (KshitijLakhani) [reply transformer_engine/pytorch/attention/__init__.py]: I think that makes a lot of sense if we want users to be able to use `RotaryPositionEmbedding` and `InferenceParams` by just importing the` transformer_engine.pytorch `module - it would make it easier for the user.
Importing `transformer_engine.pytorch` or importing `transformer_engine.pytorch.attention` will allow users to import `DPA, MHA, IP and RoPE` modules.

Added as part of latest commit SHA 968661f6ea351c3bc38bba35f4fca5cd59a30a20

File: transformer_engine/pytorch/transformer.py (hunk 3 above)
Comment (KshitijLakhani) [reply transformer_engine/pytorch/transformer.py]: Yes, had some missing tests hence not pushed those changes.
They seem duplicate from my testing and hence removed int SHA 968661f6ea351c3bc38bba35f4fca5cd59a30a20

File: transformer_engine/pytorch/attention/rope.py [hunk 5]
```diff
@@ -311,7 +310,7 @@ def apply_rotary_pos_emb(
             qkv_formats:            "thd", "bshd", "sbhd"
             context parallelism:    no
             start_positions:        yes
-            interleaving:           yes
+            interleaving            yes
```
Comment (cyanguwa) [inline transformer_engine/pytorch/attention/rope.py:313]: Why are we removing the colon here? The other ones still have the colon.

**Review by cyanguwa — APPROVED** [review]
Please address my nitpick comments and fix lint. The B200 L0 failure is due to the cuDNN version in that container. Once you update to the latest nightly, the two mask-related failures will go away.

File: transformer_engine/pytorch/attention/rope.py (hunk 5 above)
Comment (KshitijLakhani) [reply transformer_engine/pytorch/attention/rope.py]: Seems like it was introduced byme incorrectly when trying to fix a merge conflict during rebase from main
Fixed it
Thanks for catching it

File: docs/examples/te_llama/te_llama.py [hunk 6]
```diff
@@ -8,11 +8,9 @@
 from contextlib import contextmanager
 
 import torch
-from torch import nn
 
 import transformer_engine as te
-from transformer_engine.pytorch.dot_product_attention.rope import RotaryPositionEmbedding
-from transformer_engine.pytorch.fp8 import fp8_model_init
+from transformer_engine.pytorch.attention.rope import RotaryPositionEmbedding
```
Comment (KshitijLakhani) [inline docs/examples/te_llama/te_llama.py:13]: @cyanguwa same question as before: should we revert this changes as we are anyways exporting the module as before ?
Or should this change be there

File: tests/pytorch/fused_attn/test_fused_attn.py [hunk 7]
```diff
@@ -15,26 +12,22 @@
...
     FlashAttentionUtils,
     get_attention_backend,
     check_set_window_size,
     AttentionParams,
 )
-from transformer_engine.pytorch.dot_product_attention.inference import InferenceParams
-from transformer_engine.pytorch.dot_product_attention.rope import RotaryPositionEmbedding
-from transformer_engine.pytorch.constants import TE_DType
+from transformer_engine.pytorch.attention.inference import InferenceParams
+from transformer_engine.pytorch.attention.rope import RotaryPositionEmbedding
```
Comment (KshitijLakhani) [inline tests/pytorch/fused_attn/test_fused_attn.py:27]: @cyanguwa same question here too : should we change this to `from transformer_engine.pytorch.attention import (InferenceParams, RotaryPositionEmbedding)`
Or should this change be there

File: tests/pytorch/fused_attn/test_kv_cache.py [hunk 8]
```diff
@@ -11,27 +11,31 @@
...
 from transformer_engine.pytorch import make_graphed_callables
 from transformer_engine.common import recipe
 from transformer_engine.pytorch import fp8_autocast, fp8_model_init
 from transformer_engine.pytorch.transformer import (
     TransformerLayer,
 )
-from transformer_engine.pytorch.attention import DotProductAttention
-from transformer_engine.pytorch.dot_product_attention.inference import InferenceParams
-from transformer_engine.pytorch.dot_product_attention.utils import FlashAttentionUtils as fa_utils
+from transformer_engine.pytorch.attention.dot_product_attention.dot_product_attention import (
```
Comment (KshitijLakhani) [inline tests/pytorch/fused_attn/test_kv_cache.py:27]: @cyanguwa same question here too

File: transformer_engine/pytorch/__init__.py [hunk 9]
```diff
@@ -88,9 +88,10 @@ def _load_library():
 from transformer_engine.pytorch.module import Fp8Padding, Fp8Unpadding
 from transformer_engine.pytorch.module import initialize_ub
 from transformer_engine.pytorch.module import destroy_ub
-from transformer_engine.pytorch.attention import DotProductAttention
-from transformer_engine.pytorch.attention import MultiheadAttention
-from transformer_engine.pytorch.dot_product_attention.inference import InferenceParams
+from transformer_engine.pytorch.attention.dot_product_attention import DotProductAttention
+from transformer_engine.pytorch.attention.multi_head_attention import MultiheadAttention
+from transformer_engine.pytorch.attention.inference import InferenceParams
+from transformer_engine.pytorch.attention.rope import RotaryPositionEmbedding
```
Comment (KshitijLakhani) [inline transformer_engine/pytorch/__init__.py:94]: @cyanguwa same question here too

File: transformer_engine/pytorch/attention/multi_head_attention.py [hunk 10]
```diff
@@ -0,0 +1,833 @@
...
+    AttnBiasTypes,
+    dist_group_type,
+)
+from transformer_engine.pytorch.distributed import (
+    get_distributed_world_size,
+    get_distributed_rank,
+)
+
+from transformer_engine.pytorch.attention.dot_product_attention import DotProductAttention
+from transformer_engine.pytorch.attention.inference import InferenceParams
```
Comment (KshitijLakhani) [inline transformer_engine/pytorch/attention/multi_head_attention.py:31]: @cyanguwa same question here too

**Review by cyanguwa — APPROVED** [review]
